File: CS229/PS2/lr_debug.py
from __future__ import division # only import division, because older Python's / symbol could be floor division
import numpy as np # import numpy library as np, easier to type

try: # xrange functionally same as range, except xrange returns xrange object, range returns list.
    xrange
except NameError: # if cannot use xrange, use range
    xrange = range

def add_intercept(X_):
    m, n = X_.shape
    X = np.zeros((m, n + 1))
    X[:, 0] = 1
    X[:, 1:] = X_
    return X

def load_data(filename):
    D = np.loadtxt(filename)
    Y = D[:, 0]
    X = D[:, 1:]
    return add_intercept(X), Y

def calc_grad(X, Y, theta):
    m, n = X.shape
    grad = np.zeros(theta.shape)
    # the formulae below is diff cus sigmoid is from 0 to 1 but data label is -1 and 1.
    margins = Y * X.dot(theta)
    probs = 1. / (1 + np.exp(margins))
    grad = -(1./m) * (X.T.dot(probs * Y))

    return grad

def logistic_regression(X, Y):
    m, n = X.shape
    theta = np.zeros(n)
    learning_rate = 10

    i = 0
    while True:
        i += 1
        prev_theta = theta
        grad = calc_grad(X, Y, theta)
        theta = theta  - learning_rate * (grad)
        if i % 10000 == 0:
            print('Finished %d iterations' % i)
        if np.linalg.norm(prev_theta - theta) < 1e-15:
            print('Converged in %d iterations' % i)
            break
    return

def main():
    print('==== Training model on data set A ====')
    Xa, Ya = load_data('data_a.txt')
    logistic_regression(Xa, Ya)

    print('\n==== Training model on data set B ====')
    Xb, Yb = load_data('data_b.txt')
    logistic_regression(Xb, Yb)

    return

if __name__ == '__main__':
    main()

File: CS229/PS2/lr_debug.m
% compute gradient given X, y, and theta
compute_grad = @(X, y, theta) (-1 / size(X, 1)) * X' * (y ./(1 + exp((X * theta).* y)));

%% training for dataset A
disp('============= Training model on dataset A ============');
data = load('data_a.txt');
n = size(data, 1);
% adding intercept to data X
X = [ones(n, 1), data(:, 2:end)];
y = data(:, 1);

% training a logistic model
theta = zeros(size(X, 2), 1);
learning_rate = 10;
for i = 1:10^9
    prev_theta = theta;
    grad = compute_grad(X, y, theta);
    theta = theta - learning_rate * grad;
    if mod(i, 10000) == 0
        fprintf('finished iteration %d \n', i);    
    end
    if norm(theta - prev_theta) < 10^-15
        fprintf('converged in %d iterations \n', i-1);
        break;
    end
end

%% training for dataset B
disp('============= Training model on dataset A ============');
data = load('data_b.txt');
n = size(data, 1);
% adding intercept to data X
X = [ones(n, 1), data(:, 2:end)];
y = data(:, 1);

% training a logistic model
theta = zeros(size(X, 2), 1);
learning_rate = 10;
for i = 1:10^9
    prev_theta = theta;
    grad = compute_grad(X, y, theta);
    theta = theta - learning_rate * grad;
    if mod(i, 10000) == 0
        fprintf('finished iteration %d \n', i);    
    end
    if norm(theta - prev_theta) < 10^-15
        fprintf('converged in %d iterations \n', i-1);
        break;
    end
end
